python/ray/data/datasource/path_util.py

```diff
@@ -9,6 +9,111 @@
+    """
+    try:
+        import fsspec  # noqa: F401
+        from fsspec.implementations.http import HTTPFileSystem
+    except ModuleNotFoundError:
+        raise ImportError(
+            "Please install fsspec to read files from HTTP."
+        ) from None
+
+    from pyarrow.fs import FSSpecHandler, PyFileSystem
+    return PyFileSystem(FSSpecHandler(HTTPFileSystem()))
+
+
+def _wrap_fsspec_filesystem(
+    filesystem: "pyarrow.fs.FileSystem",
+) -> "pyarrow.fs.PyFileSystem":
+    """Validate and wrap an fsspec filesystem in PyArrow PyFileSystem.
+
+    Args:
+        filesystem: Filesystem to validate and wrap.
+
+    Returns:
+        PyFileSystem wrapping the fsspec filesystem.
+
+    Raises:
+        TypeError: If filesystem is not a valid pyarrow or fsspec filesystem.
+    """
+    from pyarrow.fs import FSSpecHandler, PyFileSystem
+
+    err_msg = (
+        f"The filesystem passed must either conform to "
+        f"pyarrow.fs.FileSystem, or "
+        f"fsspec.spec.AbstractFileSystem. The provided "
+        f"filesystem was: {filesystem}"
+    )
+    try:
+        import fsspec  # noqa: F401
+    except ModuleNotFoundError:
+        # If filesystem is not a pyarrow filesystem and fsspec isn't
+        # installed, then filesystem is neither a pyarrow filesystem nor
+        # an fsspec filesystem, so we raise a TypeError.
+        raise TypeError(err_msg) from None
+
+    if not isinstance(filesystem, fsspec.spec.AbstractFileSystem):
+        raise TypeError(err_msg)
+
+    return PyFileSystem(FSSpecHandler(filesystem))
+
+
+def _try_resolve_with_encoding(
+    path: str,
+    filesystem: Optional["pyarrow.fs.FileSystem"],
+) -> Tuple["pyarrow.fs.FileSystem", str]:
+    """Try resolving a path with URL encoding for special characters.
+
+    This handles paths with special characters like ';', '?', '#' that
+    may cause URI parsing errors.
+
+    Args:
+        path: The path to resolve.
+        filesystem: Optional filesystem to validate against.
+
+    Returns:
+        Tuple of (resolved_filesystem, resolved_path).
+
+    Raises:
+        Exception: Re-raises any errors that aren't URI parsing related.
+    """
+    from pyarrow.fs import _resolve_filesystem_and_path
+
+    resolved_filesystem, resolved_path = _resolve_filesystem_and_path(
+        _encode_url(path), filesystem
+    )
+    resolved_path = _decode_url(resolved_path)
+    return resolved_filesystem, resolved_path
+
+
+def _try_resolve_with_http_fsspec(
+    path: str,
+) -> Tuple["pyarrow.fs.PyFileSystem", str]:
+    """Try resolving an HTTP(S) path using fsspec HTTPFileSystem.
+
+    This is a fallback for when PyArrow doesn't recognize the HTTP(S) scheme.
+
+    Args:
+        path: The HTTP(S) path to resolve.
+
+    Returns:
+        Tuple of (http_filesystem, path).
+
+    Raises:
+        ImportError: If fsspec is not installed.
+    """
+    http_filesystem = _get_fsspec_http_filesystem()
+    return http_filesystem, path
```

Comment:
Nit: This abstraction seems like a shallow wrapper around `_get_fsspec_http_filesystem`. Directly calling `_get_fsspec_http_filesystem` might be clearer because it avoids a layer on indirection